### ISSUE TITLE
Licensing

### ISSUE THREAD

**Issue by mannol** [shahen94/react-native-video-processing#78]
AFAIK this library is using: https://github.com/kesha-antonov/ffmpeg-android-java which is licensed under GPLv3. There should be a notice in the README about the limitations of using this library.

**Comment by shahen94** [comment]
Hi @mannol . Thanks you for taking a look such things. I'm not sure that i understand what's the problem, can you please change it and open PR

**Comment by mannol** [comment]
Before I do so, it's important to understand the limitations of using GPL in a project: "(TL;DR from wikipedia) ... if GPLed parts are used in a program (and the program is distributed), then all other source code of the program needs to be made available under the same license terms". Therefore, this module (react-native-video-processing) can only be used in opensource software, unless, the license exception is made by the original author of the software in question (this is the original project that is GPL'd: https://github.com/WritingMinds/ffmpeg-android-java).

In summary: either keep using ffmpeg-android-java and be **opensource only** module or switch back to **mp4parser** to support proprietary use.

**Comment by shahen94** [comment]
I'll not switch back to mp4parser, since it don't have features which can give us ffmpeg.

So seems like we need change the licence, or i need to write my own ffmpeg-android 🤒  library from scratch with MIT licence. Thanks for letting me know about this issue.

I'll keep this thread open, and will close this when i'll decide what to do with this issue

**Comment by kesha-antonov** [comment]
Hey guys! @mannol @shahen94 

Please clarify.

If we use ```ffmpeg``` in ```react-native-video-processing``` then Can we use ```react-native-video-processing``` in commercial (not open sourced projects)?

**Comment by kesha-antonov** [comment]
Or the problem is not in the ```ffmpeg```? https://video.stackexchange.com/questions/14802/can-i-use-ffmpeg-in-a-commercial-product

Problem in ```ffmpeg-android-java```. Correct? https://github.com/WritingMinds/ffmpeg-android-java/blob/master/LICENSE.GPLv3

**Comment by shahen94** [comment]
Problem is in `ffmpeg-android-java` . We can use this library in private projects,  as i understand.

**Comment by kesha-antonov** [comment]
How about Commercial and available to public (not open sourced)?

**Comment by shahen94** [comment]
It should be fine, just this library(and forks) can't be developed in private repositories.
But i'd like to see what will say @mannol

**Comment by kesha-antonov** [comment]
If I understand correctly.

1) If we use ```ffmpeg-android-java``` -> then ```react-native-video-processing``` must be re-licensed under GPLv3.
2) If we use ```react-native-video-processing``` (which will be with GPLv3 license) -> then that Commercial project must be licensed under GPLv3 too.

Correct? @mannol @shahen94

**Comment by kesha-antonov** [comment]
If that is correct I'm going to write code that allow ```react-native-video-processing``` to work without ```ffmpeg-android-java```...

**Comment by shahen94** [comment]
@kesha-antonov how you're going to do that ?

**Comment by kesha-antonov** [comment]
I'm not sure now.
But I think it's necessarily. I propose to join our forces and make it

It's just a binary ```ffmpeg``` that needs to be executed. Seems like not so difficult task

**Comment by kesha-antonov** [comment]
I'll try to do some dummy implementation. And then we all we'll do something better

Now I'm trying to build ffmpeg with https://github.com/cisco/openh264 instead of x264. It's free. For x264 we have to pay for license. 🤕 

I'll return once I'll make progress with this. Maybe in other issue/PR

**Comment by mannol** [comment]
@kesha-antonov yes, you are correct: `ffmpeg-android-java` is the problematic code. If you remove that, you can use ffmpeg in your code as long as you comply with the ffmpeg's [guide for using ffmpeg in commercial applications](https://www.ffmpeg.org/legal.html). AFAIK `ffmpeg-android-java` is **convenience wrapper** over ffmpeg which _simplifies instalation of ffmpeg for android applications_. Creating something like it should be relatively trivial.

**Comment by kesha-antonov** [comment]
Alright.

When we have no choice. We have to do those 2 tasks I mentioned above:
1. Compile ffmpeg with openh264 (instead of libx264)
2. Put those binaries in ```react-native-video-processing``` and do dummy implementation of calling ffmpeg.

**Comment by kesha-antonov** [comment]
```2. Use dynamic linking (on windows, this means linking to dlls) for linking with FFmpeg libraries.```

What this means to us?
https://www.ffmpeg.org/legal.html

**Comment by kesha-antonov** [comment]
```13. Remove any prohibition of reverse engineering from your EULA.```

Is it OK for us?

**Comment by CaptAhak** [comment]
Hi, since GPUImage is used on the iOS Side, could the following be a viable substitute for ffmpeg alltogether? 
https://github.com/CyberAgent/android-gpuimage

Just wanted to throw the idea out there.

**Comment by kesha-antonov** [comment]
@CaptAhak AFAIK GPUImage is not used on iOS side for crop, trim, compress video.
It's just stubbed there for future work. (or maybe for filters?). I'm using only API.

Correct? @shahen94

**Comment by shahen94** [comment]
@kesha-antonov that's correct, GPUImage is still not used on iOS side for crop, we'll use it for filters

**Comment by shahen94** [comment]
@CaptAhak i'm not sure that GPUImage have the same features which have ffmpeg

**Comment by kesha-antonov** [comment]
The question is - What to use in android.

I was able to build ffmpeg with openh264. I will try to link it and use without ```ffmpeg-android-java```

**Comment by kesha-antonov** [comment]
If there any other solution for android with Trim/Crop/Compress with Permissive License I will help to integrate it here.

**Comment by mannol** [comment]
@kesha-antonov I think ffmpeg is your best bet. About the questions you asked:
> Use dynamic linking (on windows, this means linking to dlls) for linking with FFmpeg libraries

That means that you are forced to link dynamically if you were to link with FFmpeg in your binary. Considering that you will **not** link FFmpeg in your binary (or any binary using react-native-video-processing) in any way but, rather, use the FFmpeg stand-alone binary, you are covered concerning that rule.
> Remove any prohibition of reverse engineering from your EULA

Same as above. You are not linking with FFmpeg henceforth, different rules apply. See [this](https://softwareengineering.stackexchange.com/a/289800) answer for more detailed info. Specifically these lines:
> If the program uses fork and exec to invoke plug-ins, then the plug-ins are separate programs, so the license for the main program makes no requirements for them

> non-GPL application calls a GPL application at the command line -> definitely legal

Hope these answers help.

**Comment by kesha-antonov** [comment]
@mannol 
Nice. I think it's good for us.

You meant ```LGPL``` application, right? Not ```GPL```?

----
Now about binaries.

Correct me if I'm wrong:

we must have 2 binaries separately
1) App with ```react-native-video-processing```
2) ffmpeg binaries

We put binaries in ```react-native-video-processing``` in the ```assets``` directory. 
And then call them from App with ```loadFfmpeg```

Where ```loadFfmpeg``` is our new dummy method to load ffmpeg binaries (copy/execute...?)

Or what is the correct way to separate binaries to run ffmpeg correctly (with dynamic linking)?

I'm pretty new to android dev. I appreciate any help/suggestions

**Comment by shahen94** [comment]
We can put binaries in `libraries/ffmpeg` director - Then we should call `loadFFMPEG` in `RNVideoProcessingPackage.java` file

**Comment by mannol** [comment]
LGPL is more permissive than GPL. Regardless, same rules apply. About the implementation: you will, most likely, have to use [Runtime.exec()](http://docs.oracle.com/javase/6/docs/api/java/lang/Runtime.html#exec(java.lang.String[],%20java.lang.String[],%20java.io.File)) function from java standard library to execute FFmpeg process.

Note about linking: linking several libraries in a single entity creates a single program. Since you will be using FFmpeg **as a program**, **there is no linking** of FFmpeg with react-native-video-processing. Therefore, as long as you use it like that (calling Runtime.exec() on FFmpeg binary), you are set.

**Comment by kesha-antonov** [comment]
OK! Thanks guys ! 👍 
I'll try do it and will get back with questions

**Comment by mannol** [comment]
@kesha-antonov check the edit

**Comment by kesha-antonov** [comment]
Thanks! @mannol

**Comment by kesha-antonov** [comment]
Hey guys.

I've built ffmpeg with openh264.
When I try to compress video with libopenh264 it gives me error: ```CANNOT LINK EXECUTABLE. library "libopenh264.so" not found page record for 0x..... was not found (block_size=16)```

Does this mean that i should put ```libopenh264.so``` in the same directory with ```ffmpeg```?
I did not found options to disable shared libs. 

https://github.com/cisco/openh264
https://github.com/cisco/openh264/issues/1529

**Comment by kesha-antonov** [comment]
I saw that libx264 was compiled with these options and worked without errors.
```
--enable-static \
--disable-shared \
```

But I don't see these options for building openh264

Maybe I should put some other flags?

**Comment by shahen94** [comment]
Seems like you should put `libopenh264` in the project.
https://trac.pjsip.org/repos/ticket/1758

**Comment by mannol** [comment]
Shared libraries should go in `libs/ ` directory (probably `android/app/src/libs/armeabi-v7a/` for react native). I think it should work even if you place it in the same directory as your binary.

**Comment by kesha-antonov** [comment]
Alright! I'll try both

**Comment by kesha-antonov** [comment]
Same error.

https://github.com/kesha-antonov/ffmpeg-android-java/commit/de38625d61804bb23e252636cbaa4e86b8267f9e

Any thoughts?

**Comment by kesha-antonov** [comment]
Tried to put in ```node_modules/react-native-video-processing/android/src/main/jniLibs/```.
Same error

**Comment by mannol** [comment]
Try to place them in `modules/react-native-video-processing/android/src/main/jniLibs/armeabi-v7a/`

**Comment by kesha-antonov** [comment]
Already.
```
ls node_modules/react-native-video-processing/android/src/main/jniLibs
armeabi-v7a/ x86/
```

Can you help me to integrate it into ```RN-video-processing``` ?
I have ffmpeg with openh264 + libopenh264.so
It's time for task №2 - somehow integrate it here without ffmpeg-android-java

**Comment by kesha-antonov** [comment]
Maybe it copies ffmpeg to some other directory where ffmpeg can't see ```*.so``` ?
https://github.com/kesha-antonov/ffmpeg-android-java/blob/master/FFmpegAndroid/src/main/java/com/github/hiteshsondhi88/libffmpeg/FFmpegLoadLibraryAsyncTask.java#L27

**Comment by kesha-antonov** [comment]
Guys. Really need your help
Working on this > 24 hours already

Here's results:
1) Scripts to build ffmpeg with openh264: https://github.com/kesha-antonov/ffmpeg-android/commits/release/v0.4
2) Builds: https://github.com/kesha-antonov/ffmpeg-android-java/commits/master (ffmpeg with openh264 + .so files)
3) RN-video-processing with openh264: https://github.com/kesha-antonov/react-native-video-processing/commits/get_rid_of_ffmpeg_android_java

I don't have time anymore for now
Let's continue and finish it. We all need this ASAP

**Comment by kesha-antonov** [comment]
Here's what ffmpeg gives about license:

<img width="255" alt="screen shot 2017-07-12 at 12 54 05" src="https://user-images.githubusercontent.com/11584712/28122426-d6c14146-6726-11e7-9d3f-b743232973f4.png">

**Comment by kesha-antonov** [comment]
Crop and Trim are working.
Compress (which requires openh264) does not.

So maybe better to do integration and then fix that error with shared lib.

**Comment by kesha-antonov** [comment]
One more thing.

```RN-video-processing``` uses https://github.com/wseemann/FFmpegMediaMetadataRetriever which uses ```ffmpeg``` https://github.com/wseemann/FFmpegMediaMetadataRetriever/tree/master/gradle/fmmr-library/library/src/main/ffmpeg 

Right?
Should we do some about it too?

**Comment by shahen94** [comment]
I'm not sure, maybe @mannol can help you with that question.

@kesha-antonov i'm in vacation in these days, i'll try to help when i'll get to my computer

**Comment by mannol** [comment]
Looks like `FFmpegMediaMetadataRetriever` uses FFmpeg directly (links with FFmpeg), and I've checked their FFmpeg build flags which seem to comply with guidelines provided by FFmpeg project. It's ok to use it for this project.

**Comment by kesha-antonov** [comment]
Ok guys! Thanks!

**Comment by kesha-antonov** [comment]
Hello @shahen94 
Have you had time to try to link ffmpeg and "libopenh264.so" shared lib?

**Comment by shahen94** [comment]
Hi @kesha-antonov , sorry for slow reply.
I'm in vacation and will get back in the end of August. I'll work on this right away when i'll back to home

**Comment by kesha-antonov** [comment]
@shahen94 Ok!

I think we need to add warning in the README. To people know about problem with licensing

**Comment by mannol** [comment]
Getting back to this issue: I solved it by using [these](https://github.com/WritingMinds/ffmpeg-android/releases/tag/v0.3.4) prebuilt binaries for ffmpeg. Just place the armeabi-v7a ffmpeg binary in assets/ directory and you will be able to run it on your android device (to make it work on emulator, choose x86 binary instead). Now, the remaining part was to just code binary execution. There are a bunch of examples online on how to accomplish that.

Sorry I can't supply any code but, I hope I helped!

**Comment by kesha-antonov** [comment]
@mannol but these ffmpeg binaries compiled with flag --enable-gpl

And these guys https://www.ffmpeg.org/legal.html suggest to not use this flag

What do you think?

**Comment by mannol** [comment]
It's ok, using a binary executable is not the same as linking to a library. More info can be found [here](https://softwareengineering.stackexchange.com/a/289800). Please note the GPL part:
> If the program uses fork and exec to invoke plug-ins, then the plug-ins are separate programs, so the license for the main program makes no requirements for them.

**Comment by kesha-antonov** [comment]
Oh come on!
Are you sure?

This way I can use what I've built for ffmpeg-android-java 👍

**Comment by mannol** [comment]
Yup, I'm positive!

**Comment by kesha-antonov** [comment]
@mannol Cool! That's good news. Then I'll do implementation with old builds of ffmpeg 👍

**Comment by mannol** [comment]
@kesha-antonov Great! Hope to see it in **master** soon!

**Comment by kesha-antonov** [comment]
@mannol What do you think about using libx264? I think it's not free for commercial use, right? https://licensing.x264.org/en/

So if I'll include build of ffmpeg in this lib, then some people should buy license from x264, right?

**Comment by mannol** [comment]
Buying licence for libx264 is required only if the user wants to use non-GPL version of the library. The code for libx264 as used in ffmpeg is covered by GPL as you can see in their official git repository [here](http://git.videolan.org/?p=x264.git;a=blob;f=COPYING;h=d60c31a97a544b53039088d14fe9114583c0efc3;hb=HEAD).

**Comment by kesha-antonov** [comment]
@mannol Thanks. That's good.

**Comment by kesha-antonov** [comment]
Hello @mannol 
Almost done.

What do you think, can we use ```ProcessBuilder``` instead of ```Runtime.getRuntime().exec```
It seams easier to use
https://stackoverflow.com/questions/10723346/why-should-avoid-using-runtime-exec-in-java

**Comment by mannol** [comment]
@kesha-antonov Glad to hear that. Yes you can use ProcessBuilder, it does the same as exec().

**Comment by kesha-antonov** [comment]
Ok. Got it working.

Last thing. What of these https://www.ffmpeg.org/legal.html we should do (and include in README) to safely use ffmpeg? @mannol @shahen94

**Comment by kesha-antonov** [comment]
And second. I included only ```ffmpeg``` build for ```armeabi-v7a```. Do we need to include ```x86``` build or it's a legacy?

**Comment by mannol** [comment]
@kesha-antonov Great news! README should say that this project is using ffmpeg and point to the ffmpeg legal page. I think that's about it concerning legality of using ffmpeg.
>  I included only ffmpeg build for armeabi-v7a. Do we need to include x86 build or it's a legacy?

x86 is required, for example, android emulators running x86 android build. It is not necessary to be included in the build but maybe it would be nice to point out that the build cannot be used on x86 android emulators in the README.

**Comment by mradzinski** [comment]
Sorry to bump this, but still getting ```CANNOT LINK EXECUTABLE "xxx": library "libopenh264.so" not found```. @kesha-antonov, did you include `libopenh264.so` for x86 builds after all? I actually see it under the `lib` folder of my project when running it on an emulator, but for some reason the app keeps complaining it's not there.

<img width="260" alt="screen shot 2017-10-02 at 2 08 20 am" src="https://user-images.githubusercontent.com/4214209/31064855-c2916f22-a716-11e7-8041-dbb3f4eb9b07.png">

**Comment by kesha-antonov** [comment]
No, I did not include binary for x86

**Comment by mradzinski** [comment]
@kesha-antonov, I've also been trying on two ARM physical devices and I get the same exception, so I guess it's not a binary architecture issue but something related to the location of the `libopenh264.so` file. Any idea what that might be?

**Comment by kesha-antonov** [comment]
I don't use emulators on android so can't help here. On real device it works OK

**Comment by mradzinski** [comment]
I actually just tested it on a Nexus 6 and on a Samsung Galaxy 7 (both real devices, one sporting Android N and the other Android O) and it failed on both to locate `libopenh264.so`.

**Comment by kesha-antonov** [comment]
You sure you use code from master?

**Comment by kesha-antonov** [comment]
Also did you checked how to setup Android from README?
```How to setup Library```
